### PR TITLE
Add Post  "Create Universal Apps Using Next.js"

### PR DESCRIPTION
**Pull request by HeeL**
Together with Andreas Geier (@functino) we are working on the new blogpost about Next.js. This PR is very raw and I open it only for easier collaboration.

### CODE REVIEW

**Comment by HeeL** [comment]
Yeah, great idea, nextjs would be a good follow up to that. Include this
into one Post would be an overhead

Am Fr., 11. Jan. 2019, 09:27 hat Andreas <notifications@github.com>
geschrieben:

**Comment by jankowiakdawid** [comment]
Hi guys. I've read the English part of your post and I find few missing pisses, but I also noticed that there wasn't much update on it in some time.
So I just wanted to ask if you need any help on the post?

My team had Next app running in HC staging for few months and I also regularly observe how Next evolves.

**Comment by HeeL** [comment]
@jankowiakdawid thank you for your willingness to support us with this blogpost. As you've noticed, its still WIP so there are many points that are missing. I would still like to get in touch with you to discuss this topic.

File: _posts/2019-01-04-nextjs.md [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+        }
+        return (<div>
+            <button onClick={() => this.setState({ showHotel: true })}>Show Hotel</button>
+        </div>)
+    }
+}
+```
+
+Do you see what changed? We're wrapping our dynamic import with `dynamic()`. This will create a react component that will dynamically import the needed code in the momement where we actually try to render the component.
+Sounds a bit weird at first ant the benefits are not immediately obvious. But `dynamic` has a second `options` param.
```
Comment (jankowiakdawid) [inline _posts/2019-01-04-nextjs.md]:
```suggestion
Sounds a bit weird at first and the benefits are not immediately obvious. But `dynamic` has a second `options` param.
```

File: _posts/2019-01-04-nextjs.md [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+}
+```
+While this looks alomst like using a normal style tag it has some benefits: It will only apply the styles to exactly this file. So no other `div` will get a red color.
+
+And if you don't like styled-jsx you can always switch to sass, inline-css, css-in-js with [https://github.com/zeit/next.js/#importing-css--sass--less--stylus-files](just a little configuration).
+
+## Wrapping up
+
+Next.js is a powerful tool for building single page application. As it follows convention over configuration principle, it makes it look easy to do this. Projects build with Next.js mostly have the same or very similar structure. Obvious use case where you definitely have to consider using Next.js would be building prototypes. It will be also handy for teams who don't want to maintain their own webpack config and setups, but still have advantages that framework provides.
+Next.js is by all means production ready and would fit for enterprise application and not only some small projects. It is built in a way that enables you to access lower level parts when needed, which makes it extremely flexible in what it does.
```
Comment (jankowiakdawid) [inline _posts/2019-01-04-nextjs.md]: I think it would be good to add a link to showcase gallery here, so the reads could see for themselves who and for what projects uses NextJS
https://nextjs.org/showcase/

File: _posts/2019-01-04-nextjs.md [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+date: 2019-03-18 10:00:00 +0200
+categories: javascript
+author_names: ["Andreas Geier", "Sergii Paryzhskyi"]
+read_time : 10
+show_related_posts: false
+square_related:
+feature_image: posts/2019-01-04-nextjs/poster.png
+---
+
+Let's talk at first about libraries for creating universal applications. On Github you can find a large number of projects implementing the idea of universal web applications. However, all these projects have common drawbacks:
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: First sentence, `at` can be dropped.
Github -> GitHub 💅

File: _posts/2019-01-04-nextjs.md [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+
+- missing functionality such as server side rendering (SSR) or i.e. hot reloading
+- these are projects for quick start, not a library
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: You can drop `these are projects`, as this is already in context, and simply go with: `they are designed for quick start, ...`.

File: _posts/2019-01-04-nextjs.md [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+- small number of project contributors
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: Similar to above - `small number of contributors`

File: _posts/2019-01-04-nextjs.md [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+- projects were not updated when new versions of react.js were released
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: And here: `they were not ...`.
Question is - why react.js? I know it is connected to React, but it gets here all of a sudden.
Could it just go with `they were not updated when new versions of dependencies were released`?

File: _posts/2019-01-04-nextjs.md [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+
+The first successful solution was the Next.js library, which, has 3600 contributors and over 33K “stars” on Github. To evaluate the benefits of this library, consider what kind of functionality you need to provide for the operation of a universal web application:
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: You write about solution - but there is no clear problem defined anywhere. There is just a mention about common drawbacks, which are technically not a problem. How about presenting one? E.g. you wanted to create something, but most solutions were not good candidates because of reasons (drawbacks) <-- here you have a problem defined.

Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: How about linking Next.js to GitHub and adding `(...) which has (at the time of writing) 3600 (...)`?

Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: `The first ...` - were there any other? Perhaps worth mentioning?
Also, you address the reader `consider what kind`, as if he already knew the what and the why, or having the same problem. I would rephrase it to your story what you needed, what you wanted. And if so, follow the same narrative later.

File: _posts/2019-01-04-nextjs.md [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+
+- Asynchronous data loading. In order for the results of asynchronous requests to be received prior to server rendering, Next.js implements a special type of the page component, which has an asynchronous static async getInitialProps ({req}) life cycle event.
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: `static async getInitialProps ({req})` <-- to have it looking as a code example.

File: _posts/2019-01-04-nextjs.md [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+
+- Passing server component state to client.
+
+- Creating a component on the side of a web browser and linking it to an HTML document.
+
+- Routing on the server and on the client should also be universal. That is, the same definition of routing should work for both server and client code.
+
+- Code splitting. For each page, only the necessary JavaScript code should be loaded, not the entire application. When moving to the next page, the missing code should be reloaded - without reloading the same modules, and without unnecessary modules.
+
+All these tasks are successfully solved by the Next.js library. This library is based on a very simple idea. It is proposed to introduce a new type of component - “page”, in which there is an async static method async getInitialProps ({req}). The “page” component is a normal React component. This type of component can be thought of as a new type in the series: “component”, “container”, “page”.
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: It is proposed --> It introduces a new ...

Comment (smolak) [reply _posts/2019-01-04-nextjs.md]: There is this lifecycle event again. Could be just mentioned again or presented in the code or even have its own paragraph if it is important.

File: _posts/2019-01-04-nextjs.md [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+
+## Sample Project
+
+Let's create a simple project with Next.js to demonstrate how it works. You will be surprised how little we need to do in order to start our actual application. Some complexity is hidden by the library and the rule convention over configuration is helping to keep code on our side really slim.
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: You will be ... how little **you**

Comment (smolak) [reply _posts/2019-01-04-nextjs.md]: How about linking to https://en.wikipedia.org/wiki/Convention_over_configuration ?

Comment (smolak) [reply _posts/2019-01-04-nextjs.md]: (..) to keep **the** code (...) 💅

File: _posts/2019-01-04-nextjs.md [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+
+Create a new directory for our sample project `mkdir sample-app && cd sample-app`. Generate an empty npm project here with `npm init -y`. Then install Next.js together with React dependencies: `npm i next react react-dom --save`.
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: Our -> your
Also check any further occurrences.

File: _posts/2019-01-04-nextjs.md [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+
+Now we just have to add a new npm run script to our `package.json`:
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: we -> you

I mean ... assuming you are writing to your reader, and not yourselves ;)

File: _posts/2019-01-04-nextjs.md [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+
+```js
+  "scripts": {
+    "dev": "next"
+  }
+```
+
+Let's try: `npm run dev`. Unfortunately you'll see an error message in your terminal. But with  next.js, error messages are usually pretty good and tell you exactly what you have to do next:
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: (...) good and **will** tell you (...)

File: _posts/2019-01-04-nextjs.md [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+
+`Couldn't find a `pages` directory. Please create one under the project root`
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: You have a ` in a ` string. I guess you'd want to replace it with ' (inside).

File: _posts/2019-01-04-nextjs.md [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+## Showing hotel data
+
+We want to show data about an hotel. So let's create the necessary files for that:
+
+For simplicities sake we will just use a static data file for the hotel data - src/hotel.js:
+```
+module.exports = {
+    id: '1aa4c4ad-f9ea-3367-a163-8a3a6884d450',
+    name: 'Dana Beach Resort',
+    description: 'Das modern ausgestattete 5-Sterne-Hotel Dana Beach Resort liegt direkt am eigenen Sandstrand, erbaut in einem einzigartigen Stil, mit hohem Komfort, mit bestem Service und einem umfangreichen Sport- & Freizeitangebot.'
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: I would not go with German description ;) [Lorem ipsum](https://www.lipsum.com/) would also do fine here.

File: _posts/2019-01-04-nextjs.md [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+
+`Couldn't find a `pages` directory. Please create one under the project root`
+
+So go ahead, create a `pages` directory and rerun `npm run dev`. It will tell you that it is now running on port 3000 - so open localhost:3000 in your browser. It's rendering a 404 page however because we did not yet create any "page".
+
+<img src="/img/posts/2019-01-04-nextjs/404.png" alt="404 Not Found" class="centered" />
+
+Lets create an `index.js` file with the following content:
+
+`export default () => <div>Hello World</div>`
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: Missing comma! ;) Commas save lives! Example: `Let's eat grandma.` vs `Let's eat, grandma.` ;)

(Here and in other examples also)

File: _posts/2019-01-04-nextjs.md [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+- we automatically can use react and `jsx` without any webpack setup
+- if you change the text from `Hello World` to `Hello World!` it will automatically be refreshed in your browser
+- ...
+
+
+But lets focus on the "application" we want to build.
+
+## Showing hotel data
+
+We want to show data about an hotel. So let's create the necessary files for that:
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: **a** hotel

File: _posts/2019-01-04-nextjs.md [hunk 18]
```diff
@@ -0,0 +1,290 @@
...
+
+For simplicities sake we will just use a static data file for the hotel data - src/hotel.js:
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: Simplicity

File: _posts/2019-01-04-nextjs.md [hunk 19]
```diff
@@ -0,0 +1,290 @@
...
+import hotel from '../src/hotel';
+
+export default () => {
+    return (<div>
+        {hotel.name}
+    </div>)
+}
+```
+
+Check in your browser: It now shows us our hotel's name.
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: `:` -> `.`.
`:` indicates an upcoming list or an example. Here it is just one event after another 💅

File: _posts/2019-01-04-nextjs.md [hunk 20]
```diff
@@ -0,0 +1,290 @@
...
+
+## Using another route and a button
+
+What if we want to initially NOT show the data - we only want to reveal it after a click? Well that's easy and we could solve it by adding another page to show the hotel data and link it form the index.js page:
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: What if we **don't** want to show the data initially (will sound better)

File: _posts/2019-01-04-nextjs.md [hunk 21]
```diff
@@ -0,0 +1,290 @@
...
+import Router from 'next/router'
+
+export default () => {
+    return (<div>
+        <button onClick={() => Router.push('/hotel')}>Show Hotel</button>
+    </div>)
+}
+```
+
+Refresh your browser and click the button to go to the /hotel page. Again this looks like very little - but there is a bunch of stuff we get for free.
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: And here is where you should end with `;` 💅

File: _posts/2019-01-04-nextjs.md [hunk 22]
```diff
@@ -0,0 +1,290 @@
...
+
+- we see our routing in action now. And additionally to [links](https://github.com/zeit/next.js/#with-link) we can also imperatively use Router.push
+- if you open your webdeveloper toolbar's network tab you'll see that this uses client side navigation - there is no full reload of our application
+- and again in the webdeveloper toolbar you'll see that we get "bundle splitting" for free: Each page has its own JavaScript bundle - resulting in awesomely fast pages.
+
+Sure it is possible to achieve this with webpack or other bundlers manually as well - but with nextjs it is 0 effort.
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: 0 -> zero

File: _posts/2019-01-04-nextjs.md [hunk 23]
```diff
@@ -0,0 +1,290 @@
...
+            return <div>{hotel.name}</div>
+        }
+        return (<div>
+            <button onClick={() => this.showHotel()}>Show Hotel</button>
+        </div>)
+    }
+}
+```
+
+But this has one downside: Regardless if the user really looked at the hotel data - he has to load it in his browser. And our PO doesn't like that because it slows down the page for ALL users even when only 1% of them actually click on the "Show Hotel"-button. For our little example that's not much of a difference but in real world applications it sometimes is a huge improvement to only load what's really needed.
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: Regardless -> **r**egardless 💅

File: _posts/2019-01-04-nextjs.md [hunk 24]
```diff
@@ -0,0 +1,290 @@
...
+
+We can use dynamic imports to achieve that:
+
+## Dynamic Imports
+
+Next comes with a default webpack configuration that will make dynamic imports just work out of the box.
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: You can drop the `just`.

File: _posts/2019-01-04-nextjs.md [hunk 25]
```diff
@@ -0,0 +1,290 @@
...
+            return <div>{this.state.hotel.name}</div>
+        }
+        return (<div>
+            <button onClick={() => this.showHotel()}>Show Hotel</button>
+        </div>)
+    }
+}
+```
+
+Instead of a boolean state we will introduce an `hotel` state. When the user clicks on the button we dynamically import `src/hotel.js` - this will yield a promise and when that is fulfilled we can set the result to the state.
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: an -> a (state)

Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: Set ... **in** the state. 💅

File: _posts/2019-01-04-nextjs.md [hunk 26]
```diff
@@ -0,0 +1,290 @@
...
+            return <Hotel />
+        }
+        return (<div>
+            <button onClick={() => this.setState({ showHotel: true })}>Show Hotel</button>
+        </div>)
+    }
+}
+```
+
+Do you see what changed? We're wrapping our dynamic import with `dynamic()`. This will create a react component that will dynamically import the needed code in the momement where we actually try to render the component.
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: in the **moment when** (both words)

File: _posts/2019-01-04-nextjs.md [hunk 27]
```diff
@@ -0,0 +1,290 @@
...
+        <style jsx>{`
+            div {
+                color: red;
+            }
+        `}</style>
+        {hotel.name}
+    </div>)
+}
+```
+While this looks alomst like using a normal style tag it has some benefits: It will only apply the styles to exactly this file. So no other `div` will get a red color.
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: It will -> **it** will

File: _posts/2019-01-04-nextjs.md [hunk 28]
```diff
@@ -0,0 +1,290 @@
...
+
+And if you don't like styled-jsx you can always switch to sass, inline-css, css-in-js with [https://github.com/zeit/next.js/#importing-css--sass--less--stylus-files](just a little configuration).
+
+## Wrapping up
+
+Next.js is a powerful tool for building single page application. As it follows convention over configuration principle, it makes it look easy to do this. Projects build with Next.js mostly have the same or very similar structure. Obvious use case where you definitely have to consider using Next.js would be building prototypes. It will be also handy for teams who don't want to maintain their own webpack config and setups, but still have advantages that framework provides.
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: teams who -> teams **that** (who is for people)

File: _posts/2019-01-04-nextjs.md [hunk 29]
```diff
@@ -0,0 +1,290 @@
...
+Next.js is by all means production ready and would fit for enterprise application and not only some small projects. It is built in a way that enables you to access lower level parts when needed, which makes it extremely flexible in what it does.
+Not only is it really easy to use - it also has an amazing documentation. So far, every question we had was answered in their FAQ or through examples provided in their repository.
```
Comment (smolak) [inline _posts/2019-01-04-nextjs.md]: Link to docs would be nice :)

**Comment by smolak** [comment]
General remarks:
 - read through the whole thing, check from the birds-eye view if the article has a structure that has logical meaning and there are no out of the blue paragraphs / sections (write missing connections to them or remove them)
 - you mix `we, us` with `you, yours` - I would go with the latter, as you are addressing your users, but I leave the choice to you
 - any code examples, paths to files, etc. could be `represented as a code` for richer readability
 - Github -> GitHub, nextjs, Next -> Next.js (be consistent and use the correct name across the whole post, especially with next as it is also a normal word in the dictionary ;) )

Besides that I really like how you incrementally add stuff and resolve problems one by one. It is nice to read. Especially that users might think - ah, we're done, and there are things that they might not have foresee, like you did (or stuff that can be improved). Nice job 🥇

File: _posts/2019-01-04-nextjs.md [hunk 30]
```diff
@@ -0,0 +1,290 @@
...
+        }
+        return (<div>
+            <button onClick={() => this.setState({ showHotel: true })}>Show Hotel</button>
+        </div>)
+    }
+}
+```
+
+Do you see what changed? We're wrapping our dynamic import with `dynamic()`. This will create a react component that will dynamically import the needed code in the momement where we actually try to render the component.
+Sounds a bit weird at first ant the benefits are not immediately obvious. But `dynamic` has a second `options` param.
```
Comment (gateboss) [inline _posts/2019-01-04-nextjs.md]: ant -> and the benefits

**Comment by functino** [comment]
@gateboss @smolak @jankowiakdawid Thank you all for your comments. @HeeL is out of office but as soon as I have some time I'll address your comments :)

File: _posts/2019-01-04-nextjs.md (hunk 11 above)
Comment (functino) [reply _posts/2019-01-04-nextjs.md]: I think `our` is fine here and in the other places.
When we use `we` in this blogpost it means us authors + the reader.

File: _posts/2019-01-04-nextjs.md (hunk 16 above)
Comment (functino) [reply _posts/2019-01-04-nextjs.md]: I miss commas all the time, but in this case I think it really isn't necessary?

File: _posts/2019-01-04-nextjs.md (hunk 17 above)
Comment (functino) [reply _posts/2019-01-04-nextjs.md]: Oh crap. I always thought that `an hotel` is correct. Like `an hour`.  Turns out I was mislead because the `h` in `hotel` isn't silent. Thanks :)

File: _posts/2019-01-04-nextjs.md (hunk 19 above)
Comment (functino) [reply _posts/2019-01-04-nextjs.md]: I think using `:` once in a while even when not having a list or an example is fine. For me it is like: Hey something amazing will follow.

File: _posts/2019-01-04-nextjs.md (hunk 24 above)
Comment (functino) [reply _posts/2019-01-04-nextjs.md]: I agree that you can usually drop words like `just`, `simply` and so on. But in this case I'm on the edge. The `just` aims to highlight that it does not work out of the box without next.
But I'll just remove it ;)

File: _posts/2019-01-04-nextjs.md (hunk 16 above)
Comment (smolak) [reply _posts/2019-01-04-nextjs.md]: Well, it is the world's most recognizable code example ... ;) Up to you.

Comment (functino) [reply _posts/2019-01-04-nextjs.md]: Aaah, I think now I get it. I really thought there is a comma missing here: `Lets create an `index.js` file with the following content`.

Regarding the `Hello World`, I also have an excuse for that. The german wikipedia page states that `Hello World!` without comma started becoming more common :D (I missed the `!` however.

File: _posts/2019-01-04-nextjs.md (hunk 10 above)
Comment (HeeL) [reply _posts/2019-01-04-nextjs.md]: Yep, good idea to link that, since we don't expect everyone to know about coding by convention principle

**Comment by HeeL** [comment]
Thank you everyone for your reviews, last changes were applied and  it seems like we are ready to go 🚀
